### ISSUE TITLE
cast won't accept instance variable for `typ` parameter

### ISSUE THREAD

**Issue by kyle-query** [microsoft/pyright#3833]
**Describe the bug**

When calling `cast(self.t, ...)`, pyright shows an error at `self` which says `TypeVar "Self@Foo[A@Foo]" has no member "t"`.

**To Reproduce**

```python
from typing import Generic

A = TypeVar('A')

class Foo(Generic[A]):
    def __init__(self, t: Type[A]):
        self.t = t

    def foo(self) -> A:
        return cast(self.t, None)
```

**Expected behavior**

This code should produce the same result as the next example, which has no error:

```python
class Bar(Generic[A]):
    def __init__(self, t: Type[A]):
        self.t = t

    def bar(self) -> A:
        t = self.t
        return cast(t, None)
```

**Screenshots or Code**

See above example.

<img width="608" alt="Screen Shot 2022-08-14 at 2 44 53 PM" src="https://user-images.githubusercontent.com/108306629/184552455-ea010921-9439-4133-81ba-196bce908a53.png">

**VS Code extension or command-line**

Problem can be reproduced in neovim (+lspconfig) and on the command line

```
$ pyright --version
pyright 1.1.266
```

**Comment by erictraut** [comment]
I'm not clear on what you're trying to do. The `typing.cast` function performs no action at runtime other than returning the second argument. It is intended to be used for static type checking, and the first argument must be a valid static type expression. You're passing a dynamic value expression (`self.t`) as the first argument. Perhaps you have the two arguments swapped and you intended to cast `self.t` to a type of `None`? Or perhaps you meant to cast `None` to the type `A`?

**Comment by kyle-query** [comment]
> the first argument must be a valid static type expression

Thanks, that clarifies why it wouldn't work, and yes `cast(A, None)` was what I intended. Regarding the error message, are there cases where a TypeVar does have members?

**Comment by erictraut** [comment]
When a TypeVar is used as a runtime object, it has defined members such as `__name__` and `__covariant__`.

```python
A = TypeVar("A")

print(A.__name__)
```

I this particular case, `self` isn't really a `TypeVar` object at runtime. However, its inferred type is `Self`, which is modeled as a TypeVar bound to the enclosing class.
